flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/factories/HiveFunctionDefinitionFactory.java

```diff
@@ -0,0 +1,140 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.catalog.hive.factories;
+
+import org.apache.flink.api.java.typeutils.GenericTypeInfo;
+import org.apache.flink.connectors.hive.HiveTableFactory;
+import org.apache.flink.table.api.TableException;
+import org.apache.flink.table.catalog.CatalogFunction;
+import org.apache.flink.table.catalog.config.CatalogConfig;
+import org.apache.flink.table.catalog.hive.client.HiveShim;
+import org.apache.flink.table.catalog.hive.client.HiveShimLoader;
+import org.apache.flink.table.catalog.hive.descriptors.HiveCatalogValidator;
+import org.apache.flink.table.factories.FunctionDefinitionFactory;
+import org.apache.flink.table.functions.FunctionDefinitionUtil;
+import org.apache.flink.table.functions.AggregateFunctionDefinition;
+import org.apache.flink.table.functions.FunctionDefinition;
+import org.apache.flink.table.functions.ScalarFunctionDefinition;
+import org.apache.flink.table.functions.TableFunctionDefinition;
+import org.apache.flink.table.functions.hive.HiveFunctionWrapper;
+import org.apache.flink.table.functions.hive.HiveGenericUDAF;
+import org.apache.flink.table.functions.hive.HiveGenericUDF;
+import org.apache.flink.table.functions.hive.HiveGenericUDTF;
+import org.apache.flink.table.functions.hive.HiveSimpleUDF;
+import org.apache.flink.types.Row;
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.ql.exec.UDAF;
+import org.apache.hadoop.hive.ql.exec.UDF;
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDAFEvaluator;
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDAFResolver2;
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDF;
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDTF;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A factory to instantiate Hive UDFs as Flink UDFs.
+ */
+public class HiveFunctionDefinitionFactory implements FunctionDefinitionFactory {
+	private static final Logger LOG = LoggerFactory.getLogger(HiveTableFactory.class);
+
+	private final String hiveVersion;
+	private final HiveShim hiveShim;
+
+	public HiveFunctionDefinitionFactory(HiveConf hiveConf) {
+		// this has to come from hiveConf, otherwise we may lose what user specifies in the yaml file
+		this.hiveVersion = checkNotNull(hiveConf.get(HiveCatalogValidator.CATALOG_HIVE_VERSION),
+			"Hive version is not defined");
+		this.hiveShim = HiveShimLoader.loadHiveShim(hiveVersion);
+	}
+
+	@Override
+	public FunctionDefinition createFunctionDefinition(String name, CatalogFunction catalogFunction) {
```

Comment:
We might want to assert that the input type is HiveCatalogFunction.

Reply:
there's no HiveCatalogFunction. CatalogFunction has only one impl - CatalogFunctionImpl

Reply:
Yeah, that's the case today, but in the future there might be other implementations. If Hive catalog can only handle CatalogFunctionImpl.

Reply:
I'm leaning towards not doing that now because we don't assert on any catalog table/view/partition/db objects currently.

If it turns out we need to go that route, we'd better update them all at once in the future.

Reply:
ok. Maybe we can create a JIRA for this?

Reply:
sure, I created https://issues.apache.org/jira/browse/FLINK-14458 for tracking purpose